### PR TITLE
fix #114

### LINKED ISSUE #114: GenericSetup upgrade does not work in cluster environments.
### What I did:

Plone cluster with 9 instances. Run upgrade steps or profiles.

### What I expect to happen:

Profiles or upgrades are applied

### What actually happened:

They are not. 

### What version of Python and Zope/Addons I am using:

Python 3.9.8 Zope 5 latest, Plone

### Analysis

GenericSetup uses keys generated by hash(). Since hash is only the same inside one running Python each instance has its own set of keys. In a cluster the keys of one instance are used in the form and processed by another instance with different keys it can not look up.
This is probably a Python3 change on hash

### Solution 

Use a stable hashing algorithm to build the hash.

### PR DESCRIPTION
**Pull request by jensens**
see #114

### CODE REVIEW

File: src/Products/GenericSetup/upgrade.py [hunk 1]
```diff
@@ -123,7 +124,9 @@ class UpgradeEntity(object):
     """
     def __init__(self, title, profile, source, dest, desc, checker=None,
                  sortkey=0):
-        self.id = str(abs(hash('%s%s%s%s' % (title, source, dest, sortkey))))
+        id_base = f'{title}{source}{dest}{sortkey}'
```
Comment (mauritsvanrees) [inline src/Products/GenericSetup/upgrade.py:127]: GenericSetup master still supports Python 2.7 and 3.5, so we cannot use f-strings.

**Review by mauritsvanrees — CHANGES_REQUESTED** [review]
I have not seen this go wrong yet myself, but I can imagine it. I confirm that the original hash is stable on Python 2 but not Python 3. Your new hash is stable on both.
This seems a good change.

It contains f-strings, which does not work on all Python versions we support. So this should be changed.

I don't know why the GitHub Actions don't get triggered. I think the [yaml config](https://github.com/zopefoundation/Products.GenericSetup/blob/master/.github/workflows/tests.yml) is fine. Did anything change in GitHub Actions recently?

File: src/Products/GenericSetup/upgrade.py (hunk 1 above)
Comment (jensens) [reply src/Products/GenericSetup/upgrade.py:127]: ops, ok, what a pity

Comment (jensens) [reply src/Products/GenericSetup/upgrade.py]: ok, fixed that

**Review by mauritsvanrees — APPROVED** [review]

**Comment by mauritsvanrees** [comment]
@dataflake Can you re-enable the GitHub Actions? The actions don't run anymore for PRs, probably because it is too long ago that a commit was done. When I try to re-run all jobs, I get an error: "Unable to re-run disabled workflow."
I am guessing that you have more powers than I.

**Comment by dataflake** [comment]
I reenabled them.

**Comment by mauritsvanrees** [comment]
Thanks. All checks are green now.

**Comment by jensens** [comment]
Is this package part of the rule "merge yourself if positive review"?

**Comment by mauritsvanrees** [comment]
> Is this package part of the rule "merge yourself if positive review"?

Yes.

**Pull request merged**